SU2_CFD/include/numerics/turbulent/turb_sources.hpp

```diff
@@ -28,331 +26,567 @@
 #include "../scalar/scalar_sources.hpp"
 
 /*!
- * \class CSourcePieceWise_TurbSA
+ * \class CSAVariables
+ * \brief Structure with SA common auxiliary functions and constants.
+ */
+struct CSAVariables {
+  /*--- List of constants ---*/
+  const su2double cv1_3 = pow(7.1, 3), k2 = pow(0.41, 2), cb1 = 0.1355, cw2 = 0.3, ct3 = 1.2, ct4 = 0.5,
+                  cw3_6 = pow(2, 6), sigma = 2.0 / 3.0, cb2 = 0.622, cb2_sigma = cb2 / sigma,
+                  cw1 = cb1 / k2 + (1 + cb2) / sigma, cr1 = 0.5;
+
+  /*--- List of auxiliary functions ---*/
+  su2double ft2, d_ft2, r, d_r, g, d_g, glim, fw, d_fw, Ji, d_Ji, S, Shat, d_Shat, fv1, d_fv1, fv2, d_fv2;
+
+  /*--- List of helpers ---*/
+  su2double Omega, dist_i_2, inv_k2_d2, inv_Shat, g_6, norm2_Grad, gamma_bc;
+};
+
+/*!
+ * \class CSourceBase_TurbSA
  * \brief Class for integrating the source terms of the Spalart-Allmaras turbulence model equation.
- * \ingroup SourceDiscr
- * \author A. Bueno.
+ * The variables that are subject to change in each variation/correction have their own class.
+ * \note Additional source terms (e.g. compressibility) are implemented as decorators.
  */
+template <class FlowIndices, class Omega, class ft2, class ModVort, class rFunc, class SourceTerms>
 class CSourceBase_TurbSA : public CNumerics {
-protected:
-  su2double cv1_3;
-  su2double k2;
-  su2double cb1;
-  su2double cw2;
-  su2double ct3;
-  su2double ct4;
-  su2double cw3_6;
-  su2double cb2_sigma;
-  su2double sigma;
-  su2double cb2;
-  su2double cw1;
-  su2double cr1;
-
+ protected:
   su2double Gamma_BC = 0.0;
-  su2double intermittency;
-  su2double Production, Destruction, CrossProduction;
 
+  /*--- Residual and Jacobian ---*/
   su2double Residual, *Jacobian_i;
-private:
-  su2double Jacobian_Buffer; /// Static storage for the Jacobian (which needs to be pointer for return type).
+  su2double Jacobian_Buffer; /*!< \brief Static storage for the Jacobian (which needs to be pointer for return type). */
 
-protected:
-  const bool incompressible = false, rotating_frame = false;
-  bool roughwall = false;
+  const FlowIndices idx; /*!< \brief Object to manage the access to the flow primitives. */
+  const bool rotating_frame = false;
+  const bool transition = false;
 
-public:
+ public:
   /*!
    * \brief Constructor of the class.
-   * \param[in] val_nDim - Number of dimensions of the problem.
-   * \param[in] val_nVar - Number of variables of the problem.
+   * \param[in] nDim - Number of dimensions of the problem.
    * \param[in] config - Definition of the particular problem.
    */
-  CSourceBase_TurbSA(unsigned short val_nDim, unsigned short val_nVar, const CConfig* config);
+  CSourceBase_TurbSA(unsigned short nDim, unsigned short, const CConfig* config)
+      : CNumerics(nDim, 1, config),
+        idx(nDim, config->GetnSpecies()),
+        rotating_frame(config->GetRotating_Frame()),
+        transition(config->GetKind_Trans_Model() == TURB_TRANS_MODEL::BC) {
```

Comment:
I also did not know but the LM model works in combination with SST, SA-LM and SST-BC do not "exist".

Reply:
It seems that other transition models for SA might be implemented in a near future, see #1496. So, it would be a friendly correction to modify that line.

I will add a TODO annotation in the code.

Reply:
> I will add a TODO annotation in the code.

:disappointed: